xtm/xtm_api.h

```diff
@@ -0,0 +1,222 @@
+ * @param[in] queue - xtm_queue to get file descriptor.
+ * @retval    xtm queue file descriptor for consumer thread.
+ */
+int
+xtm_queue_consumer_fd(struct xtm_queue *queue);
+
+/**
+ * Return file descriptor, that should be watched by producer thread to
+ * become readable. When it became readable, producer may push to queue,
+ * being sure that there is a place in it (there is a little race between
+ * consumer and producer thread, so there is a little chance, that this
+ * descriptor become readable, but there is no free space in the queue. In
+ * this case producer thread must poll this descriptor again.
+ * @param[in] queue - xtm_queue to get file descriptor.
+ * @retval    xtm queue file descriptor for producer thread.
+ */
+int
+xtm_queue_producer_fd(struct xtm_queue *queue);
+
+/**
+ * Calls all functions contained in the queue.
+ * In case when xtm_queue was created with XTM_QUEUE_PRODUCER_NEED_NOTIFICATIONS
+ * flag, user should check, that queue was full and producer thread failed to
+ * put to the queue (using xtm_queue_was_full, after this function), and if it
+ * was so, notify producer thread (see READMY.md for simple example).
+ * @param[in] queue - xtm_queue.
+ * @retval    count of invoked functions if success.
+ */
+unsigned
+xtm_queue_invoke_funs_all(struct xtm_queue *queue);
+
+/**
+ * Push pointer to the queue. This function does not notify the consumer
+ * thread, but only push to the queue. To notify the consumer thread you
+ * must call xtm_queue_notify_consumer. The less often you notify, the
+ * greater the performance, but the greater the latency.
+ * @param[in] queue - xtm_queue to push.
+ * @param[in] ptr   - pointer to push.
+ * @retval    0 if queue has space. Otherwise -1 with errno set to ENOBUFS.
+ */
+int
+xtm_queue_push_ptr(struct xtm_queue *queue, void *ptr);
+
+/**
+ * Get up to count elements from queue and save them in pointer array.
+ * In case when xtm_queue was created with XTM_QUEUE_PRODUCER_NEED_NOTIFICATIONS
+ * flag, user should check, that queue was full and producer thread failed to
+ * put to the queue (using xtm_queue_was_full, after this function), and if it
+ * was so, notify producer thread (see READMY.md for simple example).
+ * @param[in]  queue           - xtm_queue containing pointers.
+ * @param[out] ptr_array       - pointer array to save pointers.
+ * @param[in]  ptr_array_count - maximum count of pointers, that can
+ *                               be extracted.
+ * @retval     count of extracted elements.
+ */
+unsigned
+xtm_queue_pop_ptrs(struct xtm_queue *queue, void **ptr_array,
+		   unsigned ptr_array_count);
+
+/**
+ * Read data from queue pipe.
+ * @param[in] fd - file descriptor to consume.
+ * @retval    0 on success. Otherwise -1 with errno set appropriately
+ *            (as in read(2), since it implies read from internal fd).
+ */
+int
+xtm_queue_consume(int fd);
+
+/**
+ * Check that xtm queue was full, and producer thread
+ * failed to put to the queue.
```

Comment:
You definitely should document the side effect. It may also be a good idea to rename this function to something like "get_and_reset_queue_full_flag()". Code readability is a must.

Reply:
I specifically didn't specify how calling this function affects the internals of the queue.

Reply:
Right now the name is misleading - functions with such names typically have no side effects. Error-prone.

Reply:
Indeed, the functions called 'is', 'was', 'has' etc never should have side effects. However a long name which is also inconsistent without code style, like the one void234 proposed, is also bad. I would propose `xtm_queue_get_reset_was_full()` or `xtm_queue_pop_was_full()`.

Reply:
Fixed to `xtm_queue_get_reset_was_full`

Reply:
The side effect is still not documented.

Reply:
Add description. Fixed